### ISSUE TITLE
Wrong values on interactivy

### ISSUE THREAD

**Issue by alasarr** [CartoDB/carto-vl#685]
Hello,

It looks the feature's properties returned using an interactivity are not right. 

This is our test:
1. Create a `featureEnter` interactivity.
2. When the event is received we log: coordinates, zoom and take the first feature.
3. Create a quick script in Python to fetch/decode the same tile. We're using coordinates and zoom from step 2. 
4. **The feature of step 2 doesn't have the same properties than step 3**.  

Probably there's some kind of wrong reference when using a dataset with lots of attributes (~400).

CARTO VL script:

```html
<!DOCTYPE html>
<html>
<head>
  <title>Filter based styling | CARTO</title>
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <meta charset="UTF-8">
  <!-- Include CARTO VL JS -->
  <script src="../../dist/carto-vl.js"></script>
  <!-- Include Mapbox GL JS -->
  <script src="https://libs.cartocdn.com/mapbox-gl/v0.45.0-carto1/mapbox-gl.js"></script>
  <!-- Include Mapbox GL CSS -->
  <link href="https://libs.cartocdn.com/mapbox-gl/v0.45.0-carto1/mapbox-gl.css" rel="stylesheet" />
  <link rel="stylesheet" type="text/css" href="../style.css">
</head>
<body>
  <div id="map"></div>
  <aside class="toolbox">
    <div class="box">
      <header>
        <h1>Filter based styling</h1>
      </header>
      <section>
        <p class="description open-sans">Filter data in the browser</p>
      </section>
      <footer class="js-footer"></footer>
    </div>
  </aside>
  <div id="loader">
    <div class="CDB-LoaderIcon CDB-LoaderIcon--big">
      <svg class="CDB-LoaderIcon-spinner" viewBox="0 0 50 50">
        <circle class="CDB-LoaderIcon-path" cx="25" cy="25" r="20" fill="none"></circle>
      </svg>
    </div>
  </div>
  <script>
    const map = new mapboxgl.Map({
      container: 'map',
      style: 'https://basemaps.cartocdn.com/gl/dark-matter-gl-style/style.json',
      center: [-73.99, 40.73],
      zoom: 11,
      scrollZoom: true,
      dragRotate: false,
      touchZoomRotate: false,
    });

    // Define user
    carto.setDefaultAuth({
      user: 'alasarr',
      apiKey: 'default_public'
    });

    var xhttp = new XMLHttpRequest();
    xhttp.onreadystatechange = function() {
        if (this.readyState == 4 && this.status == 200) {
           const data = JSON.parse(xhttp.responseText),
           uniqueid = 'id',
           properties = {},
           fields = data['vector_layers'][0].fields
           ;

           for (const key of Object.keys(fields)) {
            properties[key] = {
              'type': (key === uniqueid && fields[key].toLowerCase() === 'string') ? 'category' : fields[key].toLowerCase()
            };
          }

          const source = new carto.source.MVT(data['tiles'][0], {
            properties: properties,
            idProperty: uniqueid
          }, 'usa_02');

          const viz = new carto.Viz(`
            @id: $id
            @test: $__02_01_2012__apparel__growth_country_score
            strokeWidth: 1,
            strokeColor: red
          `);

          const layer = new carto.Layer('layer', source, viz);
          layer.addTo(map, 'watername_ocean');
          layer.on('loaded', hideLoader);

          const interactivity = new carto.Interactivity([layer]);

          interactivity.on('featureEnter', (event) => {
            console.log(`Coordinates: ${event.coordinates}`)
            console.log(`id: ${event.features[0].variables.id.value}`)
            console.log(`__02_01_2012__apparel__growth_country_score: ${event.features[0].variables.test.value}`)
            console.log(`zoom: ${map.getZoom()}`);
          });

        }
    };
    xhttp.open("GET", "https://<tiles>", true);
    xhttp.send();

    function hideLoader() {
      document.getElementById('loader').style.opacity = '0';
    }
  </script>
</body>
</html>
```

Console log output:
```
Coordinates: LngLat(-74.01403259278754, 40.75445125164754)
id: 36061009900
__02_01_2012__apparel__growth_country_score: 729 => WRONG!
zoom: 11
```

Python script (it uses as input previous console.log):
```python
import requests
import mercantile
def getFeatureProperties(long, lat, zoom, id, prop=None):
    x, y, z = mercantile.tile(long, lat, zoom)
    url = 'http://<server>/{z}/{x}/{y}.pbf'.format(z = z, x = x, y = y)
    print(url)
    r = requests.get(url, stream=True)
    if r.status_code == 200:
        decoded_data = mapbox_vector_tile.decode(r.content)
        for f in decoded_data['usa_02']['features']:
            if f['properties']['id']==id:
                return f['properties'].get(prop, None) if prop else f['properties']
        return None
getFeatureProperties(-74.01403259278754, 40.75445125164754, 11, '36061009900','__02_01_2012__apparel__growth_country_score')
``` 
Python script output:
```
http://<server>/11/602/769.pbf
'46' => Right!
```

**Comment by alasarr** [comment]
cc: @rochoa

**Comment by davidmanzanares** [comment]
Hi @alasarr,

Can you check that the `metadata` provided to the `carto.source.MVT` is valid, I think you could be passing `type: string` in some cases and `string` is not a valid type.

In fact, if I had to make a guess I would remove this code: `key === uniqueid && `, could you check that too?

**Comment by padawannn** [comment]
We have replaced `'type': (key === uniqueid && fields[key].toLowerCase() === 'string') ? 'category' : fields[key].toLowerCase()` by `'type': (fields[key].toLowerCase() === 'string') ? 'category' : fields[key].toLowerCase()` and it's still not working

**Comment by alasarr** [comment]
@davidmanzanares it happens because our tile had integer encoded as string. We've changed it and it's working.

Wrong behaviour:
```json
{
"__02_01_2012__apparel__growth_country_score": "800"
}
```

Valid: 
```json
{
"__02_01_2012__apparel__growth_country_score": 800
}
```

I think it's a bug should be fixed, but since we've a workaround, we can not a priority for us right now.

If you need our tileset to debug, just let me know and I can send you a MBTile with the number encoded as string.

**Comment by rochoa** [comment]
All this starts when specifying the type of a property as a `number` but it is actually a `category`.

The property decoder finds that the real type is a `category`/`string` and tries to internally-encode it, categorizing it with a number.

https://github.com/CartoDB/carto-vl/blob/101ec5aec2675548da240a4781c4e01aef51b47c/src/sources/Mvt.js#L226-L228

https://github.com/CartoDB/carto-vl/blob/101ec5aec2675548da240a4781c4e01aef51b47c/src/renderer/Metadata.js#L25-L36

When that happens, we already lost track of the original number—remember, it was a cateogry, not a number—because it will get an auto-incremental number that references the original category/string.

However, later we retrieve that supposed number, we never check for the real type again and we trust the metadata type, so we don't translate the auto-incremental/generated internal number back to the original category because we only know that is a number.

https://github.com/CartoDB/carto-vl/blob/101ec5aec2675548da240a4781c4e01aef51b47c/src/renderer/Dataframe.js#L278-L293

I guess the best alternative here is to throw an error while decoding in `decodeProperty` if there is a mismatch between the metadata type and the actual type.

**Comment by davidmanzanares** [comment]
Fixed by #742
